### PR TITLE
fix(cli): prevent crash when displaying incomplete test workflow executions

### LINKED ISSUE #6891: CLI Panic with Unhelpful Error When API Returns Incomplete Execution Object
**Environment:**

Testkube CLI version: 2.3.0

Testkube Helm Chart version: 2.1.236

Kubernetes version: 1.32 on AWS EKS

**Description:**

When the Testkube API returns an incomplete TestWorkflowExecution object (e.g., due to connection issues, wrong endpoint protocol, or API errors), the CLI crashes with an unhelpful segmentation fault instead of a user-friendly error message.

**Steps to Reproduce:**

Configure Testkube CLI to use HTTP endpoint instead of HTTPS (or any scenario where API returns incomplete data, this one worked in my case)

Run: testkube run tw test-name --watch

CLI panics with segmentation fault

**Actual Output:**

```
Test Workflow Execution:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1e0e14e]

goroutine 1 [running]:
github.com/kubeshop/testkube/cmd/kubectl-testkube/commands/testworkflows/renderer.printPrettyOutput(, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, 0x0, 0x0, {0x0, ...}, ...})
/home/runner/work/testkube/testkube/cmd/kubectl-testkube/commands/testworkflows/renderer/testworkflowexecution_obj.go:53 +0x4e
github.com/kubeshop/testkube/cmd/kubectl-testkube/commands/testworkflows/renderer.PrintTestWorkflowExecution(, {_, _}, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, 0x0, ...})
/home/runner/work/testkube/testkube/cmd/kubectl-testkube/commands/testworkflows/renderer/testworkflowexecution_obj.go:26 +0x265
github.com/kubeshop/testkube/cmd/kubectl-testkube/commands/testworkflows.processExecutions(0xc000b86f08, {0xc00089ef20, 0x1, 0x0?}, {0x31bd728, 0xc00032d0e0}, 0x1, {0x1, {0x0, 0x0}, ...}, ...)
/home/runner/work/testkube/testkube/cmd/kubectl-testkube/commands/testworkflows/run.go:432 +0x1a7
github.com/kubeshop/testkube/cmd/kubectl-testkube/commands/testworkflows.NewRunTestWorkflowCmd.runTestWorkflow.func1(0xc000b86f08, {0xc000075780, 0x1, 0x4})
/home/runner/work/testkube/testkube/cmd/kubectl-testkube/commands/testworkflows/run.go:217 +0x6be
github.com/spf13/cobra.(*Command).execute(0xc000b86f08, {0xc000075740, 0x4, 0x4})
/home/runner/go/pkg/mod/github.com/spf13/mailto:mailto:cobra@v1.8.1/command.go:989 +0xb1b
github.com/spf13/cobra.(*Command).ExecuteC(0x49513c0)
/home/runner/go/pkg/mod/github.com/spf13/mailto:mailto:cobra@v1.8.1/command.go:1117 +0x44f
github.com/spf13/cobra.(*Command).Execute(...)
/home/runner/go/pkg/mod/github.com/spf13/mailto:mailto:cobra@v1.8.1/command.go:1041
github.com/spf13/cobra.(*Command).ExecuteContext(...)
/home/runner/go/pkg/mod/github.com/spf13/mailto:mailto:cobra@v1.8.1/command.go:1034
github.com/kubeshop/testkube/cmd/kubectl-testkube/commands.Execute()
/home/runner/work/testkube/testkube/cmd/kubectl-testkube/commands/root.go:234 +0x576
main.main()
/home/runner/work/testkube/testkube/cmd/kubectl-testkube/main.go:24 +0xf
```

**Expected Behavior:**

CLI should display a user-friendly error message such as:

```
Error: Failed to execute test workflow 'test-name'
Reason: API returned incomplete execution data (Workflow field is missing)
Hint: Check your API endpoint configuration and ensure the API server is accessible
```

**Root Cause:**

In testworkflowexecution_obj.go:53, the code directly accesses execution.Workflow.Name without checking if execution.Workflow is nil:

``` go
func printPrettyOutput(ui *ui.UI, execution testkube.TestWorkflowExecution) {
    ui.Info("Test Workflow Execution:")
    ui.Warn("Name:                ", execution.Workflow.Name)  // Line 53 - PANIC if Workflow is nil
```

**Suggested Fix:** (AI generated - not tested)

Add nil checks before accessing nested fields:

``` go
func printPrettyOutput(ui *ui.UI, execution testkube.TestWorkflowExecution) {
    ui.Info("Test Workflow Execution:")
    
    if execution.Workflow == nil {
        ui.Err(errors.New("incomplete execution data received from API - Workflow field is missing"))
        ui.Warn("Hint: Check your API endpoint configuration (HTTP vs HTTPS) and API server accessibility")
        return
    }
    
    ui.Warn("Name:                ", execution.Workflow.Name)
    // ... rest of the code
}
```

**Impact:**

Users waste time debugging cryptic segmentation faults

The actual issue (API connectivity/configuration) is hidden

Poor user experience for a common misconfiguration scenario

### PR DESCRIPTION
**Pull request by zucchinidev**
## Pull request description 
    
When API connectivity issues cause incomplete execution data to be received, the CLI would crash with an unhelpful nil pointer error, preventing users from diagnosing the underlying problem.
    
This change ensures graceful degradation by detecting missing workflow references and displaying a clear error message with actionable guidance about common causes such as API endpoint misconfiguration.
    
Users can now view available execution details even when workflow metadata is incomplete, significantly improving the troubleshooting experience.
    
Closes #6891 

## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [x] tested locally
- [x] tested on cluster
- [ ] added new dependencies
- [ ] updated the docs
- [x] added a test

## Breaking changes

- None

## Changes

 **File:** `cmd/kubectl-testkube/commands/testworkflows/renderer/testworkflowexecution_obj.go`
  - Added nil check for `execution.Workflow` before accessing its fields
  - Display clear error message when workflow metadata is missing
  - Provide actionable hint about common causes (HTTP vs HTTPS misconfiguration)
  - Continue rendering available execution details for troubleshooting

  **File:** `cmd/kubectl-testkube/commands/testworkflows/renderer/testworkflowexecution_obj_test.go` (new)
  - TestPrintPrettyOutput_NilWorkflow - Verifies no panic with nil workflow reference
  - TestPrintPrettyOutput_NilWorkflowWithInitError - Tests edge case with initialization errors
  - TestPrintPrettyOutput_ValidWorkflow - Regression test ensuring normal flow still works

## Fixes

  Issue #6891 - CLI panic when displaying test workflow executions with incomplete data

**Root Cause**

`testworkflowexecution_obj.go` accessed execution.Workflow.Name without checking if execution.Workflow was nil, causing a panic when the API returned incomplete data.

**Testing Performed**

1. Unit Tests:

```$ go test ./cmd/kubectl-testkube/commands/testworkflows/renderer -v -run TestPrintPrettyOutput
  === RUN   TestPrintPrettyOutput_NilWorkflow
  --- PASS: TestPrintPrettyOutput_NilWorkflow (0.00s)
  === RUN   TestPrintPrettyOutput_NilWorkflowWithInitError
  --- PASS: TestPrintPrettyOutput_NilWorkflowWithInitError (0.00s)
  === RUN   TestPrintPrettyOutput_ValidWorkflow
  --- PASS: TestPrintPrettyOutput_ValidWorkflow (0.00s)
  PASS
```

2. Integration Testing (Kind Cluster):
  - Installed TestWorkflow CRDs in Kubernetes cluster
  - Created TestWorkflow via CLI: `testkube create testworkflow -f panic-demo-workflow.yaml`
  - Executed workflow to generate real execution: `testkube run testworkflow panic-demo`
  - Verified execution retrieval works with valid workflow reference
  - Simulated incomplete data by setting workflow: null in MongoDB
  - Confirmed fix handles corrupted data gracefully

### Before (Panic - Buggy Code)

**Scenario**: Execution with workflow: null in database

```
  $ ./bin/app/testkube get testworkflowexecution 695c386f1b9db186b2aa8b38 --client direct

  Test Workflow Execution:
  panic: runtime error: invalid memory address or nil pointer dereference
  [signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x105f7db28]

  goroutine 1 [running]:
  github.com/kubeshop/testkube/cmd/kubectl-testkube/commands/testworkflows/renderer.printPrettyOutput(...)
      .../testworkflowexecution_obj.go:53 +0x48
  github.com/kubeshop/testkube/cmd/kubectl-testkube/commands/testworkflows/renderer.TestWorkflowExecutionRenderer(...)
      .../testworkflowexecution_obj.go:47 +0xac
  ...
```

**User Impact**: CLI crashes, no execution details shown, no guidance on how to fix.

 ### After (Graceful Error - With Fix)

**Scenario 1**: Valid execution with workflow reference
  
```
$ ./bin/app/testkube get testworkflowexecution 695c386f1b9db186b2aa8b38 --client direct

  Test Workflow Execution:
  Name:                 panic-demo
  Execution ID:         695c386f1b9db186b2aa8b38
  Execution name:       panic-demo-1
  Execution namespace:
  Execution number:     1
  Requested at:         2026-01-05 22:17:19.454 +0000 UTC
  Disabled webhooks:    false
  Status:               running
  Queued at:            2026-01-05 22:17:21 +0000 UTC
  Started at:           2026-01-05 22:17:21 +0000 UTC
```
  ✅ Normal operation - works perfectly

**Scenario 2**: Execution with workflow: null (corrupted data)
  
```
$ ./bin/app/testkube get testworkflowexecution 695c386f1b9db186b2aa8b38 --client direct

  Test Workflow Execution:

  ⨯ incomplete execution data received from API: missing Workflow field 💔
  Hint: check your API endpoint configuration (HTTP vs HTTPS) and API server accessibility
  Execution ID:         695c386f1b9db186b2aa8b38
  Execution name:       panic-demo-1
  Execution namespace:
  Execution number:     1
  Requested at:         2026-01-05 22:17:19.454 +0000 UTC
  Disabled webhooks:    false
  Status:               running
  Queued at:            2026-01-05 22:17:21 +0000 UTC
  Started at:           2026-01-05 22:17:21 +0000 UTC
```
 
✅ Graceful error handling - no panic, clear guidance, execution details still visible

**User Impact**: Users get actionable error message, can still see execution details, and know where to look for the problem

### CODE REVIEW

File: cmd/kubectl-testkube/commands/testworkflows/renderer/testworkflowexecution_obj.go [hunk 1]
```diff
@@ -50,7 +50,15 @@ func TestWorkflowExecutionRenderer(client client.Client, ui *ui.UI, obj interfac
 
 func printPrettyOutput(ui *ui.UI, execution testkube.TestWorkflowExecution) {
 	ui.Info("Test Workflow Execution:")
-	ui.Warn("Name:                ", execution.Workflow.Name)
+
+	if execution.Workflow != nil {
+		ui.Warn("Name:                ", execution.Workflow.Name)
+	} else {
+		ui.NL()
+		ui.Err(errors.New("incomplete execution data received from API: missing Workflow field"))
```
Comment (olensmar) [inline cmd/kubectl-testkube/commands/testworkflows/renderer/testworkflowexecution_obj.go:58]: could/should we output any other fields from the execution object that could be helpful in troubleshooting?

Comment (zucchinidev) [reply cmd/kubectl-testkube/commands/testworkflows/renderer/testworkflowexecution_obj.go:58]: Hi @olensmar,

Thank you for taking the time to review this PR and for the thoughtful feedback. I really appreciate the guidance.

Great question about additional troubleshooting fields. The current implementation follows a graceful degradation approach: when workflow is nil, we display all available execution metadata that doesn't depend on the workflow reference. Currently this includes:

**Currently displayed**:
- Execution identifiers (ID, name, namespace, number)
- Timing information (scheduledAt, queuedAt, startedAt, finishedAt, duration)
- Execution status and configuration (status, disableWebhooks, tags)
- Initialization errors (if present)

**Additional fields that could aid troubleshooting**:
1. runnerId - Shows which runner processed this execution
2. statusAt - When the status last changed (currently we show scheduledAt but not statusAt)
3. groupId - Helps identify correlated executions in the same batch/group
4. resolvedWorkflow.name - If workflow is nil but resolvedWorkflow exists, it could show the actual executed workflow

The logic is currently at `testworkflowexecution_obj.go:54-82` where we display fields inside the if `execution.Id != ""` block.

**Proposed Enhancement (if valuable)**

I'd propose adding these 3 specific fields when workflow is missing, as they directly answer key troubleshooting questions:

1. **`runnerId` (HIGH value)**
- **Why**: Identifies which runner had the issue
- **Troubleshooting**: "Are all failures from one runner?" → runner-specific config/connectivity problem
- **Use case**: Correlate with infrastructure logs for that specific runner

2. **`statusAt` (HIGH value)**
- **Why**: Shows when status last changed
- **Troubleshooting**: Compare timeline (scheduledAt → statusAt → queuedAt) to determine if workflow was missing from creation or became null later
- **Use case**: Correlate with API restart times or network outage windows

3. **`groupId` (MEDIUM value)**
- **Why**: If part of a batch, determines scope of problem
- **Troubleshooting**: "One execution or entire batch affected?" → different root causes
- **Use case**: Identify parallel execution issues vs isolated failures

**Implementation location**: Display these right after the error/hint (line ~60), making them prominent for troubleshooting:

```go
} else {
    ui.NL()
    ui.Err(errors.New("incomplete execution data received from API: missing Workflow field"))
    ui.Warn("Hint: check your API endpoint configuration (HTTP vs HTTPS) and API server accessibility")
    ui.NL()
    
    // Troubleshooting context
    if execution.RunnerId != "" {
        ui.Warn("Runner ID:           ", execution.RunnerId)
    }
    if !execution.StatusAt.IsZero() {
        ui.Warn("Status changed at:   ", execution.StatusAt.String())
    }
    if execution.GroupId != "" {
        ui.Warn("Group ID:            ", execution.GroupId)
    }
}
```

This keeps the change minimal (3 fields) while maximizing diagnostic value. Happy to implement this if you think it adds value, or take a different approach based on your experience with these types of issues.

**Would you like me to**:
- Proceed with this selective enhancement?
- Go with a different set of fields?
- Keep the current implementation as-is (minimal fix)?

Thanks again for the review and guidance!

**Review by vsukhin — COMMENTED** [review]
hey @zucchinidev DId you see this issue, because execution.Workflow  can't be nil by design, it's created with this fileld?
